### PR TITLE
split fully qualified name into name + module path (reported as classname)

### PR DESCRIPTION
**Pull request by MalteSchledjewski**
GitLab apparently needs a `classname`attribute.

I propose that the fully qualified name is split at the last `::` into `{classname}::{name}`

This requires bachp/junit-report-rs#5 to be merged.

### CODE REVIEW

**Comment by MalteSchledjewski** [comment]
junit-report v0.3 was released so this should build now

**Comment by johnterickson** [comment]
Sorry I missed this @MalteSchledjewski ! I am adding some more tests (https://github.com/johnterickson/cargo2junit/pull/31). Once that merges, could you merge and update tests as needed? This looks like a great contribution - I just want to be extra safe :)

File: src/main.rs [hunk 1]
```diff
@@ -101,32 +108,37 @@ fn parse<T: BufRead>(
                 SuiteEvent::Started { test_count: _ } => {
                     assert!(current_suite.is_none());
                     assert!(tests.is_empty());
-                    let mut ts = TestSuite::new(&format!("{} #{}", suite_name_prefix, suite_index));
-                    ts.set_timestamp(timestamp);
+                    let ts = TestSuite::new(&format!("{} #{}", suite_name_prefix, suite_index))
```
Comment (MalteSchledjewski) [inline src/main.rs:111]: The API of `junit-report` changed.

File: src/expected_outputs/multi_suite_success.json.out [hunk 2]
```diff
@@ -1,68 +1,32 @@
 <?xml version="1.0" encoding="utf-8"?>
 <testsuites>
   <testsuite id="0" name="cargo test #0" package="testsuite/cargo test #0" tests="5" errors="0" failures="0" hostname="localhost" timestamp="2020-06-15T20:58:06.568821800+00:00" time="0">
-    <testcase name="sharedkey::create_signature" time="0" />
-    <testcase name="client::list_tables" time="0" />
-    <testcase name="client::delete_tables" time="0" />
-    <testcase name="table::create_and_delete_table" time="0" />
-    <testcase name="blob::create_and_delete_container" time="0" />
-    <system-out />
```
Comment (MalteSchledjewski) [inline src/expected_outputs/multi_suite_success.json.out:9]: Empty `system-out` and `system-err` are now omitted by `junit-report`.

File: src/main.rs [hunk 3]
```diff
@@ -101,32 +108,37 @@ fn parse<T: BufRead>(
...
-                    .as_mut()
-                    .expect("Test event found outside of suite!");
-                let duration_ns = match (duration,exec_time) {
-                    (_,Some(s)) => {
+            Event::Test {
+                event,
+                duration,
+                exec_time,
+            } => {
+                let current_test_suite = current_suite
```
Comment (johnterickson) [inline src/main.rs:129]:
```suggestion
                let current_suite = current_suite
                    .as_mut()
                    .expect("Test event found outside of suite!");
```

File: src/main.rs [hunk 4]
```diff
@@ -137,16 +149,20 @@ fn parse<T: BufRead>(
                     }
                     TestEvent::Ok { name } => {
```
Comment (johnterickson) [inline src/main.rs:150]:
```suggestion
                    TestEvent::Ok { name } => {
                        assert!(tests.remove(&name));
                        let (name, module_path) = split_name(&name);
                        *current_suite = current_suite.clone().add_testcase(
                            TestCase::success(&name, duration).set_classname(module_path.as_str()),
                        );
                    }
                    TestEvent::Failed { name, stdout } => {
                        assert!(tests.remove(&name));
                        let (name, module_path) = split_name(&name);
                        *current_suite = current_suite.clone().add_testcase(
                            TestCase::failure(&name, duration, "cargo test", &stdout)
                                .set_classname(module_path.as_str()),
                        );
                    }
```

**Comment by johnterickson** [comment]
meh I can just make the tweaks myself :)

Thanks for your contribution!

@MalteSchledjewski Can you try this out E2E in your scenario with:
```
cargo install --git ...
```

If that looks good, I'll push to crates.io

**Comment by MalteSchledjewski** [comment]
591339af99aa1e5a6a396e04b545d27507c099a7 works like a charm

**Comment by johnterickson** [comment]
Thanks @MalteSchledjewski I just published:
https://crates.io/crates/cargo2junit

**Comment by darxriggs** [comment]
Could you set the `classname` attribute only when it is available or maybe to a default value like `default`?
Having only `src/main.rs` and `src/lib.rs` without further modules there is no `path::` prefix in the JSON output.
This results in `classname=""` in the XML output which prevents navigating the test cases in Jenkins.

**Comment by johnterickson** [comment]
@darxriggs I made an issue over here: https://github.com/johnterickson/cargo2junit/issues/34

Is this something you'd want to contribute? I use Azure DevOps Pipelines so I'm unable to have a fast inner loop in testing it out.

**Comment by darxriggs** [comment]
I am sorry to say that for the foreseeable future I have no plans to work on this.

Currently I am using a simple workaround by just removing the empty `classname` attribute from the results file like this `sed -i -e 's/classname=""//' results.xml`. Having it removed, Jenkins properly shows the results. This might easily be implemented as a test case.

**Pull request merged**